modules/heat_conduction/src/userobjects/SelfShadowSideUserObject.C

```diff
@@ -0,0 +1,412 @@
+      _triangles.push_back(Triangle{{cse.node_ref(3), cse.node_ref(1), cse.node_ref(4)}, id});
+      _triangles.push_back(Triangle{{cse.node_ref(4), cse.node_ref(2), cse.node_ref(5)}, id});
+      _triangles.push_back(Triangle{{cse.node_ref(3), cse.node_ref(4), cse.node_ref(5)}, id});
+      break;
+
+    case libMesh::QUAD4:
+      _triangles.push_back(Triangle{{cse.node_ref(0), cse.node_ref(1), cse.node_ref(2)}, id});
+      _triangles.push_back(Triangle{{cse.node_ref(2), cse.node_ref(3), cse.node_ref(0)}, id});
+      break;
+
+    case libMesh::QUAD8:
+      _triangles.push_back(Triangle{{cse.node_ref(0), cse.node_ref(4), cse.node_ref(7)}, id});
+      _triangles.push_back(Triangle{{cse.node_ref(4), cse.node_ref(1), cse.node_ref(5)}, id});
+      _triangles.push_back(Triangle{{cse.node_ref(5), cse.node_ref(2), cse.node_ref(6)}, id});
+      _triangles.push_back(Triangle{{cse.node_ref(6), cse.node_ref(3), cse.node_ref(7)}, id});
+      _triangles.push_back(Triangle{{cse.node_ref(6), cse.node_ref(7), cse.node_ref(4)}, id});
+      _triangles.push_back(Triangle{{cse.node_ref(4), cse.node_ref(5), cse.node_ref(6)}, id});
+      break;
+
+    case libMesh::QUAD9:
+      _triangles.push_back(Triangle{{cse.node_ref(0), cse.node_ref(4), cse.node_ref(7)}, id});
+      _triangles.push_back(Triangle{{cse.node_ref(4), cse.node_ref(1), cse.node_ref(5)}, id});
+      _triangles.push_back(Triangle{{cse.node_ref(5), cse.node_ref(2), cse.node_ref(6)}, id});
+      _triangles.push_back(Triangle{{cse.node_ref(6), cse.node_ref(3), cse.node_ref(7)}, id});
+      _triangles.push_back(Triangle{{cse.node_ref(8), cse.node_ref(6), cse.node_ref(7)}, id});
+      _triangles.push_back(Triangle{{cse.node_ref(8), cse.node_ref(5), cse.node_ref(6)}, id});
+      _triangles.push_back(Triangle{{cse.node_ref(8), cse.node_ref(4), cse.node_ref(5)}, id});
+      _triangles.push_back(Triangle{{cse.node_ref(8), cse.node_ref(7), cse.node_ref(4)}, id});
+      break;
+
+    default:
+      mooseError("Unsupported FACE type");
+  }
+}
+
+bool
+SelfShadowSideUserObject::check2DIllumination(const Point & qp, const SideIDType & id)
+{
+  const auto x = qp(0);
+  const auto y = qp(1);
+
+  // loop over all line segments until one is found that provides shade
+  for (const auto & line : _lines)
+  {
+    // make sure a side never shades itself
+    if (line.id == id)
+      continue;
+
+    const int imax = line.node[0](1) > line.node[1](1) ? 0 : 1;
+    const auto y1 = line.node[1 - imax](1);
+    const auto y2 = line.node[imax](1);
+
+    if (y >= y1 && y <= y2)
+    {
+      // segment is in line with the QP in radiation direction. Is it in front or behind?
+      const auto x1 = line.node[1 - imax](0);
+      const auto x2 = line.node[imax](0);
+
+      // compute intersection location
+      const auto xs = (x2 - x1) * (y - y1) / (y2 - y1) + x1;
+      // tolerance is required to avoid an element to self shadow (might be problematic w. higher
```

Comment:
tolerance is required, but where is it?

Reply:
ha, good catch. I changed the algorithm so that an element cannot cast a shadow on itself. This is means that super deformed higher order elements are not shading correctly. But this would be hard to achieve with the tolerance solution, too. At least the new approach works pretty well out of the box without the user having to worry about the tolerance. I'll remove the comment.